pkg/kubelet/cm/dra/state/state_checkpoint.go

```diff
@@ -0,0 +1,215 @@
+	sync.RWMutex
+	cache             *claimInfoCache
+	checkpointManager checkpointmanager.CheckpointManager
+	checkpointName    string
+}
+
+// NewCheckpointState creates new State for keeping track of claim info  with checkpoint backend
+func NewCheckpointState(stateDir, checkpointName string) (*stateCheckpoint, error) {
+	checkpointManager, err := checkpointmanager.NewCheckpointManager(stateDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize checkpoint manager: %v", err)
+	}
+	stateCheckpoint := &stateCheckpoint{
+		cache:             newClaimInfoCache(),
+		checkpointManager: checkpointManager,
+		checkpointName:    checkpointName,
+	}
+
+	if err := stateCheckpoint.restoreState(); err != nil {
+		//nolint:staticcheck // ST1005 user-facing error message
+		return nil, fmt.Errorf("could not restore state from checkpoint: %v, please drain this node and delete the memory manager checkpoint file %q before restarting Kubelet",
+			err, filepath.Join(stateDir, checkpointName))
+	}
+
+	return stateCheckpoint, nil
+}
+
+// restores state from a checkpoint and creates it if it doesn't exist
+func (sc *stateCheckpoint) restoreState() error {
+	sc.Lock()
+	defer sc.Unlock()
+	var err error
+
+	checkpoint := NewDRAManagerCheckpoint()
+	if err = sc.checkpointManager.GetCheckpoint(sc.checkpointName, checkpoint); err != nil {
+		if err == errors.ErrCheckpointNotFound {
+			return sc.storeState()
+		}
+		return err
+	}
+
+	for _, entry := range checkpoint.Entries {
+		//@TODO add Annotation
+		claimInfo := ClaimInfo{ClaimInfoState: ClaimInfoState{DriverName: entry.DriverName,
+			ClaimUID:   entry.ClaimUID,
+			ClaimName:  entry.ClaimName,
+			Namespace:  entry.Namespace,
+			PodUIDs:    entry.PodUIDs,
+			CdiDevices: entry.CdiDevices,
+		},
+		}
+
+		err = sc.cache.add(entry.ClaimName, entry.Namespace, &claimInfo)
+		if err != nil {
+			return err
+		}
+	}
+
+	klog.V(2).InfoS("State checkpoint: restored state from checkpoint")
+
+	return nil
+}
+
+// saves state to a checkpoint, caller is responsible for locking
+func (sc *stateCheckpoint) storeState() error {
+	checkpoint := NewDRAManagerCheckpoint()
+	checkpoint.Entries = ClaimInfoStateList{}
+	for _, infoClaim := range sc.cache.claimInfo {
+		claimInfoState := ClaimInfoState{
+			DriverName: infoClaim.DriverName,
+			ClaimUID:   infoClaim.ClaimUID,
+			ClaimName:  infoClaim.ClaimName,
+			Namespace:  infoClaim.Namespace,
+			PodUIDs:    infoClaim.PodUIDs,
+			CdiDevices: infoClaim.CdiDevices,
+		}
+		checkpoint.Entries = append(checkpoint.Entries, &claimInfoState)
+	}
+
+	err := sc.checkpointManager.CreateCheckpoint(sc.checkpointName, checkpoint)
+	if err != nil {
+		klog.ErrorS(err, "Could not save checkpoint")
+		return err
+	}
+	return nil
+}
+
+func (sc *stateCheckpoint) AddPodReference(claimName, namespace string, podUID types.UID) {
+	sc.Lock()
+	defer sc.Unlock()
+
+	claimInfo := sc.cache.get(claimName, namespace)
+	claimInfo.addPodReference(podUID)
+	err := sc.storeState()
```

Comment:
@moshe010 Can tis implicit way of storing state cause storing inconsistent state? Device manager is triggering state updates explicitly as far as I know. @klueska was it done in purpose?

Reply:
Just to make sure I understand, the concern is that the state store is too early? Do you think it better to do explicate call at the end of  PrepareResources here[1]?

[1] - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/dra/manager.go#L161

Reply:
You understood it correctly. I'm still not 100% sure we need this though. I'd like to hear your and @klueska's opinions.

Reply:
I'll take a closer look at this once my other concerns are addressed.